crypto/engine/eng_init.c

```diff
@@ -29,9 +29,8 @@ int engine_unlocked_init(ENGINE *e)
          * OK, we return a functional reference which is also a structural
          * reference.
          */
-        e->struct_ref++;
+        ENGINE_UP_STRUCT_REF(e, &ref_cnt);
         e->funct_ref++;
```

Comment:
Errr....  I'm unsure why you removed `ENGINE_UP_FUNC_REF`...  it seems odd that one counter wouldn't be treated atomically while the other would.  I *can* understand that there was confusion with your attempt at using `CRYPTO_atomic_read`, but I don't think that's reason enough to remove atomicity entirely...

Reply:
On the other hand, considering this is under global lock anyway ('cause that's how it's called), one can wonder why you need atomicity on the structural lock.

Reply:
The more I think of it, the more I'm thinking that it's a bit odd to run the init function *before* having increased the ref counts.  It's like the init function is called for an ENGINE that there's no reference to, that strikes me as odd now that I'm looking at it again.  I don't remember how we were thinking way back when, unfortunately...

Reply:
It isn't possible (at least I couldn't see how to do it) to handle the functional references using the CRYPTO_{DOWN|UP}_REF calls because the functional reference counter needs to be read without an update in several places.

I switched to using the raw atomic add and atomic read operations for the functional reference counter.  @kroeckx again objected to the presence of the wrapper macros on the read.  I attempted to re-architect the code using the per engine lock but it was problematic (i.e. more complex than I'd like, I can go into details if desired), so I reverted to using the global lock without the wrapper macros since it is a smaller change.  I don't have an issue putting the wrapper macros back but they'll not be atomic operations unless I can have a read in addition to the add.

Reply:
The structural reference count is not always updated under the global lock.  Sometimes it is updated using atomic operations and other times under the lock without atomic operations.  By not being updated in a consistent manner (the two methods don't interoperate properly), the structural reference count could go awry.  This is the entire point of this PR.

The functional reference counts was an addition.

Reply:
Running init after the increment doesn't help if it fails since there needs to be a decrement and another thread can interfere.

Reply:
Another thread *cannot* interfere, since `engine_unlocked_init` is called under the protection of the global lock (I just checked through all calls to make sure Im not lying).